### PR TITLE
Bump mypy version and fix new errors

### PR DESCRIPTION
**Pull request by patrickvalenzuela**
These errors exist on master ahead of this PR:

`awspricing/offers.py:227: error: Argument 1 to "_validate_reserved_price_args" of "EC2Offer" has incompatible type "Optional[str]"; expected "str"
awspricing/offers.py:227: error: Argument 3 to "_validate_reserved_price_args" of "EC2Offer" has incompatible type "Optional[str]"; expected "str"
awspricing/offers.py:243: error: Argument 2 to "_get_reserved_offer_term" of "EC2Offer" has incompatible type "List[Optional[str]]"; expected "List[str]"
awspricing/offers.py:306: error: Argument 1 to "_validate_reserved_price_args" of "EC2Offer" has incompatible type "Optional[str]"; expected "str"
awspricing/offers.py:306: error: Argument 3 to "_validate_reserved_price_args" of "EC2Offer" has incompatible type "Optional[str]"; expected "str"
awspricing/offers.py:322: error: Argument 2 to "_get_reserved_offer_term" of "EC2Offer" has incompatible type "List[Optional[str]]"; expected "List[str]"
awspricing/offers.py:409: error: Argument 1 to "hash_attributes" of "AWSOffer" has incompatible type "*List[Optional[str]]"; expected "str"
awspricing/offers.py:451: error: Argument 1 to "_validate_reserved_price_args" of "RDSOffer" has incompatible type "Optional[str]"; expected "str"
awspricing/offers.py:451: error: Argument 3 to "_validate_reserved_price_args" of "RDSOffer" has incompatible type "Optional[str]"; expected "str"
awspricing/offers.py:467: error: Argument 2 to "_get_reserved_offer_term" of "RDSOffer" has incompatible type "List[Optional[str]]"; expected "List[str]"
awspricing/offers.py:530: error: Argument 1 to "_validate_reserved_price_args" of "RDSOffer" has incompatible type "Optional[str]"; expected "str"
awspricing/offers.py:530: error: Argument 3 to "_validate_reserved_price_args" of "RDSOffer" has incompatible type "Optional[str]"; expected "str"
awspricing/offers.py:546: error: Argument 2 to "_get_reserved_offer_term" of "RDSOffer" has incompatible type "List[Optional[str]]"; expected "List[str]"`

### CODE REVIEW

File: awspricing/offers.py [hunk 1]
```diff
@@ -88,7 +88,7 @@ def _normalize_region(self, region):  # type: (Optional[str]) -> str
         return region
 
     @staticmethod
-    def hash_attributes(*attributes):  # type: (*str) -> str
+    def hash_attributes(*attributes):  # type: (*Any) -> str
```
Comment (patrickvalenzuela) [inline awspricing/offers.py]: I'm not sure of the right fix here

Comment (garrettheel) [reply awspricing/offers.py]: `*Optional[str]`?

Comment (patrickvalenzuela) [reply awspricing/offers.py]: `mypy --py2 awspricing
awspricing/offers.py:93: error: Incompatible return value type (got "Optional[str]", expected "str")
awspricing/offers.py:93: error: Value of type variable "AnyStr" of "join" of "str" cannot be "Optional[str]"
make: *** [test_mypy] Error 1`

Comment (garrettheel) [reply awspricing/offers.py]: You're using a newer version of `mypy` than what `mypy-requirements.txt` specifies, but here's a diff that will work with the latest version for you (please bump in reqs too):

```
diff --git a/awspricing/offers.py b/awspricing/offers.py
index f0be443..885195c 100644
--- a/awspricing/offers.py
+++ b/awspricing/offers.py
@@ -88,7 +88,7 @@ class AWSOffer(object):
         return region
 
     @staticmethod
-    def hash_attributes(*attributes):  # type: (*Any) -> str
+    def hash_attributes(*attributes):  # type: (*str) -> str
         """Generate a hash for attributes to use in reverse mappings."""
         return '|'.join(attributes)
 
@@ -226,6 +226,11 @@ class EC2Offer(AWSOffer):
         # type: (...) -> float
         self._validate_reserved_price_args(
             lease_contract_length, offering_class, purchase_option)
+
+        assert lease_contract_length is not None
+        assert offering_class is not None
+        assert purchase_option is not None
+
         sku = self.get_sku(
             instance_type,
             operating_system=operating_system,
@@ -262,7 +267,7 @@ class EC2Offer(AWSOffer):
         return hourly
 
     def _get_reserved_offer_term(self, sku, term_attributes):
-        # type: (str, List[Optional[str]]) -> Dict[str, Any]
+        # type: (str, List[str]) -> Dict[str, Any]
         term_attributes_hash = self.hash_attributes(*term_attributes)
         all_terms = self._offer_data['terms']['Reserved'][sku]
         sku_terms = self._reserved_terms_to_offer_term_code[sku]
@@ -305,6 +310,11 @@ class EC2Offer(AWSOffer):
         # type: (...) -> float
         self._validate_reserved_price_args(
             lease_contract_length, offering_class, purchase_option)
+
+        assert lease_contract_length is not None
+        assert offering_class is not None
+        assert purchase_option is not None
+
         sku = self.get_sku(
             instance_type,
             operating_system=operating_system,
@@ -383,12 +393,12 @@ class RDSOffer(AWSOffer):
         )
 
         # Lazily-loaded cache to hold offerTermCodes within a SKU
-        self._reserved_terms_to_offer_term_code = defaultdict(dict)
+        self._reserved_terms_to_offer_term_code = defaultdict(dict)  # type: Dict[str, Dict]
 
     def get_sku(self,
                 instance_type,               # type: str
                 database_engine,             # type: str
-                license_model=None,          # type: str
+                license_model=None,          # type: Optional[str]
                 deployment_option=None,      # type: Optional[str]
                 database_edition=None,       # type: Optional[str]
                 region=None                  # type: Optional[str]
@@ -396,8 +406,11 @@ class RDSOffer(AWSOffer):
         region = self._normalize_region(region)
         deployment_option = deployment_option or self.default_deployment_option
 
+        if license_model is None:
+            raise ValueError("License model is required")
+
         attributes = [instance_type, database_engine,
-                      deployment_option, license_model, region]
+                      deployment_option, license_model, region]  # type: List[str]
 
         if database_edition is not None:
             attributes.append(database_edition)
@@ -450,6 +463,11 @@ class RDSOffer(AWSOffer):
         # type: (...) -> float
         self._validate_reserved_price_args(
             lease_contract_length, offering_class, purchase_option)
+
+        assert lease_contract_length is not None
+        assert offering_class is not None
+        assert purchase_option is not None
+
         sku = self.get_sku(
             instance_type,
             database_engine,
@@ -486,7 +504,7 @@ class RDSOffer(AWSOffer):
         return hourly
 
     def _get_reserved_offer_term(self, sku, term_attributes):
-        # type: (str, List[Optional[str]]) -> Dict[str, Any]
+        # type: (str, List[str]) -> Dict[str, Any]
         term_attributes_hash = self.hash_attributes(*term_attributes)
         all_terms = self._offer_data['terms']['Reserved'][sku]
         sku_terms = self._reserved_terms_to_offer_term_code[sku]
@@ -529,6 +547,11 @@ class RDSOffer(AWSOffer):
         # type: (...) -> float
         self._validate_reserved_price_args(
             lease_contract_length, offering_class, purchase_option)
+
+        assert lease_contract_length is not None
+        assert offering_class is not None
+        assert purchase_option is not None
+
         sku = self.get_sku(
             instance_type,
             database_engine,
```

Comment (patrickvalenzuela) [reply awspricing/offers.py]: thanks for the patch 🤡

**Pull request merged**